linera-base/src/crypto/mod.rs

```diff
@@ -86,11 +88,27 @@ pub enum AccountSecretKey {
 #[derive(Eq, PartialEq, Copy, Clone, Debug, Serialize, Deserialize)]
 pub enum AccountSignature {
     /// Ed25519 signature.
-    Ed25519(ed25519::Ed25519Signature),
+    Ed25519 {
+        /// Signature of the value.
+        signature: ed25519::Ed25519Signature,
+        /// Public key of the signer.
+        public_key: ed25519::Ed25519PublicKey,
+    },
     /// secp256k1 signature.
-    Secp256k1(secp256k1::Secp256k1Signature),
+    Secp256k1 {
+        /// Signature of the value.
+        signature: secp256k1::Secp256k1Signature,
+        /// Public key of the signer.
+        public_key: secp256k1::Secp256k1PublicKey,
+    },
     /// EVM secp256k1 signature.
-    EvmSecp256k1(secp256k1::evm::EvmSignature),
+    // No public key b/c it can be recovered from the signature.
+    EvmSecp256k1 {
+        /// Signature of the value.
+        signature: secp256k1::evm::EvmSignature,
+        /// The EVM address of the signer.
+        address: [u8; 20],
```

Comment:
Do we need that? I thought in this case it's computable from `signature` alone.

Reply:
Where do we have only the signature and not the message? I'd imagine that everywhere where we care about the signer we also have the message.

I'm just wondering if we can save those 20 bytes.

It's also not great that we're adding 20 or even 32 bytes to every validator signature in every certificate now, even though we definitely don't need them there.

Reply:
Done in [Recover address for Evm signatures](https://github.com/linera-io/linera-protocol/pull/4088/commits/0f0732603c3d239a7c03e3eee1f68a25c99c3b9a)

Reply:
ok so this is where I don't necessarily agree with @afck apparently :)

Reply:
Sorry, I hadn't realized that the _validator_ signatures don't even use this type, so we won't waste any bytes there!

I don't have a strong opinion about this. I was thinking we'd not only save the 20 bytes but also make it less likely that we forget one of the two comparisons: the expected address with the `address` field here, and the `address` field with the recovered key. But I'm fine with reverting the change, too.

Reply:
Agreed to bring back the inlined address.